[roshnidaksha] iP

roshnidaksha opened the pull request.

roshnidaksha committed:
feat(level-0): rename chatbot to Planit and implement greeting and exit

roshnidaksha committed:
refactor(level-0): split main method into greet and exit methods

roshnidaksha committed:
Update README.md to match actual product name

roshnidaksha committed:
Implement Echo class to repeat user input

roshnidaksha committed:
Add playGame method to Planit class

Planit task management system has multiple features (echoing user input
and managing tasks).

Keeping all the logic in main method makes it hard to read.

Creating a separate method for playing a game improves code readability.

roshnidaksha committed:
Create Task and TaskList classes

The Task class includes attributes to define individual tasks, while
the TaskList class stores all tasks and manages related methods.

roshnidaksha committed:
Create manageTask method in Planit Class

roshnidaksha committed:
Rename TaskList.java to TaskManager.java and implement marking of tasks

Methods inside TaskList.java like addTask should not be static as it
does not affect other instances of TaskList object.

Rename TaskList.java to TaskManager.java and move static manageTask
method from Planit class to non-static run method of TaskManager class.

Update other attributes and methods in TaskManager class as non-static.

roshnidaksha committed:
Add Javadoc comments at appropriate places

roshnidaksha committed:
Create subclasses, Todo, Deadline and Event for Task class

There are different types of tasks that have some common behaviors
(e.g. toString()) and attributes (e.g. description, isDone).

Extracting the common behavior into a super class allows us to use
polymorphism.

Let's assign Task class as a parent class and create three other
subclasses. Using inheritance is preferable in this situation to
avoid code duplication.

roshnidaksha committed:
Implement task creation feature

roshnidaksha committed:
Move task-related methods from TaskManager to new TaskList class

TaskManager handles both user input and task management, violating the
single responsibility principle.

Extracting methods like addTask, displayAllTasks, and toggleTaskStatus
into TaskList improves modularity.

roshnidaksha committed:
Update testing to include task creation and mark/unmarking tasks

roshnidaksha committed:
Planit class: remove method PlayGame

roshnidaksha committed:
Add UI class for user input and message display

roshnidaksha committed:
Create CommandType enum to define task-related commands

roshnidaksha committed:
Update tests to align with new UI

ChingWhee commented on src/main/java/Planit.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Scanner;
+
+public class Planit {
```
Since "plan it" is two words, you can name the class as PlanIt (capitalize 'i')

ChingWhee commented on src/main/java/Planit.java:
```diff
@@ -0,0 +1,31 @@
+        while (!exitPlanit) {
+            Ui.showPlanitUserGuide();
+            String userChoice = Ui.getUserInput();
+            if (userChoice.equalsIgnoreCase("planit") || userChoice.equals("1")) {
```
I like how there are different modes :)

ChingWhee commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.Scanner;
+
+public class Ui {
```
Great that you have a class that stores all the graphic constants

ChingWhee commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * Takes in user input and act appropriately.
+     */
+    public void run() {
```
I like how you handle the tasks and ui logic separately

yyingg-243 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,58 @@
+public class TaskList {
+    private Task[] tasks = new Task[100];
```
I suggest to use constant such as MAX_ITEMS to avoid magic numbers

yyingg-243 commented on src/main/java/CommandType.java:
```diff
@@ -0,0 +1,6 @@
+public enum CommandType {
+    LIST,
+    MARK, UNMARK,
+    TODO, DEADLINE, EVENT,
+    BYE
+}
```
Good usage of enum class

yyingg-243 commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,128 @@
+     *
+     * @param message Error message to display.
+     */
+    public static void showError(String message) {
+        System.err.println(LINE_PREFIX + message);
+    }
+
+    /**
+     * Displays the goodbye message and exits the runtime.
+     */
+    public static void exitProgram() {
+        showToUser(PLANIT_MESSAGE_GOODBYE, DIVIDER, DIVIDER);
+        System.exit(0);
+    }
+
+    /**
+     * Prompts for the command and reads the text entered by the user.
+     *
+     * @return Full line entered by the user.
+     */
+    public static String getUserInput() {
+        System.out.print(LINE_PREFIX + "> ");
+        String inputLine = SCANNER.nextLine();
+        while (inputLine.trim().isEmpty()) {
+            inputLine = SCANNER.nextLine();
+        }
+        return inputLine;
+    }
+
+    /**
+     * Splits raw user input into command word and command arguments string.
+     *
+     * @param rawUserInput Full line entered by the user.
+     * @return size 2 array; first element is the command type and second element is the arguments string
+     */
+    public static String[] splitCommandWordAndArgs(String rawUserInput) {
+        String[] split = rawUserInput.trim().split("\\s+", 2);
+        return split.length == 2 ? split : new String[] { split[0], "" };
+    }
+}
```
Great job in naming methods as it is easy to understand

yyingg-243 commented on src/main/java/Planit.java:
```diff
@@ -0,0 +1,31 @@
+            String userChoice = Ui.getUserInput();
+            if (userChoice.equalsIgnoreCase("planit") || userChoice.equals("1")) {
+                taskManager.run();
+            } else if (userChoice.equalsIgnoreCase("play") || userChoice.equals("2")) {
+                Echo.echo();
+            } else if (userChoice.equalsIgnoreCase("bye") || userChoice.equals("3")) {
```
Could consider of naming "play", "bye", "3" as constant to improve readability

yyingg-243 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,55 @@
+            String commandArgs = commandTypeAndParams[1];
+
+            CommandType commandTypeEnum = CommandType.valueOf(commandType.toUpperCase());
+            switch (commandTypeEnum) {
+            case LIST:
+                taskList.displayAllTasks();
+                break;
+            case MARK:
+            case UNMARK:
+                int taskId = Integer.parseInt(commandArgs) - 1;
+                taskList.setTaskStatus(taskId, commandTypeEnum == CommandType.MARK);
+                break;
+            case TODO:
+            case DEADLINE:
+            case EVENT:
+                taskList.addTask(commandTypeAndParams);
+                break;
+            case BYE:
+                Ui.showTaskManagerExitMessage();
+                exitTaskManager = true;
+                break;
+            default:
+                Ui.showError("Invalid command type: " + commandType);
+            }
+            Ui.showToUser(Ui.DIVIDER);
+        }
```
Good usage of switch statement

yyingg-243 reviewed (commented):
Great job and everything looks good to me

roshnidaksha committed:
Make Task class abstract

Task has three subclasses: Todo, Deadline, and Event.
Since tasks must belong to one of these types, making Task abstract
prevents instantiation of generic tasks.

roshnidaksha committed:
Add Parser class to process user input

roshnidaksha committed:
Add error handling for invalid user input

roshnidaksha committed:
Add EmptyCommandException for incomplete arguments

roshnidaksha committed:
Merge branch 'branch-Level-5'

roshnidaksha committed:
Reorganize classes into appropriate Java packages

roshnidaksha committed:
Merge branch 'branch-A-Packages'

roshnidaksha committed:
Update text-ui-test to include error handling

roshnidaksha committed:
Use Java Collections to manage task list

roshnidaksha committed:
Implement task deletion functionality

roshnidaksha committed:
Add Storage class to save tasks in hard disk

roshnidaksha committed:
Implement saving tasks to disk

roshnidaksha committed:
Merge branch 'branch-Level-6'

roshnidaksha committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/planit/task/TaskList.java

roshnidaksha committed:
Use Java Path objects for cross-platform file path management

roshnidaksha committed:
Refactor TaskList to use a Map for task storage by type

Let's replace the single ArrayList<Task> with a
Map<String, ArrayList<Task>> to categorize tasks by type (todo,
deadline, event).

This improves task retrieval and makes it easier to display tasks
categorized by type.

roshnidaksha committed:
Display user-friendly messages and package app as JAR

- Added descriptive messages to inform users about task retrieval
status.
- Packaged the application as a JAR file for easy distribution.

roshnidaksha committed:
Extract command classes to inherit from abstract Command class

Refactored command classes to inherit from a common Command class,
ensuring each command has its own implementation.
This improves code clarity and maintainability.

roshnidaksha committed:
fix(command): ensure new command instances are created for each user input

Previously, the chatbot reused the same command instances from a static
HashMap, causing input states to persist across multiple user
interactions. This led to issues such as error messages appearing
repeatedly even after correcting the input.

To resolve this:
- Refactored the `Command.getCommand()` method to return a new
instance of the requested command each time it is called.
- Ensured that each command execution is stateless by creating fresh
instances, preventing unexpected behaviors due to shared state.

roshnidaksha committed:
fix(parser): Correctly handle input starting with a key

Previously, the `parseKeyValuePairs` function would incorrectly treat
input starting with a key (e.g., "/by tmr") as part of the description.
This led to issues where key-value pairs were not parsed properly when
no description was provided.

Let's adjust the logic to split the input correctly, treating everything
before the first `/` as the description, and the rest as key-value pairs.

roshnidaksha committed:
Extract Planit messages to messages package

Moved all chatbot response messages to a new package for better
maintainability, consistency, and scalability. Centralized message
management allows for easier updates and potential future localization.

roshnidaksha committed:
Merge branch 'branch-A-MoreOOP'

roshnidaksha committed:
Add help command to display user guide

roshnidaksha committed:
fix(parser): remove matcher.matches() check before matcher.find()

Previously, the code used matcher.matches() to validate the entire input
before processing key-value pairs with matcher.find(). However, this caused
issues when only one match existed, as matcher.find() evaluated to false.

- Removed `if (!matcher.matches())` check.
- Now, `matcher.find()` runs directly, ensuring it processes key-value pairs
  correctly even when only key-value pair is present.

roshnidaksha committed:
Add DateParser class to handle date parsing

roshnidaksha committed:
Implement date command to show tasks for a specific date

roshnidaksha committed:
fix(DateParser): correct handling of date and datetime inputs

Date-only inputs (e.g., 23/2/2025) were not correctly parsed, even when
time was optional in the formatter. This was because LocalDateTime
objects cannot store dates without a time component.

To fix this:
- First attempt parsing the input as a LocalDateTime.
- If that fails, parse it as a LocalDate and convert it to LocalDateTime
  with a default time of 00:00.

This ensures proper handling of both date-only and datetime inputs.

roshnidaksha committed:
Update runtest.bat in text UI tests to reflect new features

roshnidaksha committed:
Implement find command to show tasks with a specific keyword

roshnidaksha committed:
Merge pull request #1 from roshnidaksha/branch-Level-8

Implement Date and Time Parsing and Storage

roshnidaksha committed:
Merge pull request #2 from roshnidaksha/branch-Level-9

Implement finding tasks by keyword

roshnidaksha committed:
Add JavaDoc comments to classes and methods

roshnidaksha committed:
Improve user message formatting for better readability

roshnidaksha committed:
Add CommandResult class to encapsulate command execution results

roshnidaksha committed:
Include relevant tasks in CommandResult for list, find, and date

roshnidaksha committed:
fix(Parser): handle /on flag properly for date command

roshnidaksha committed:
Merge pull request #3 from roshnidaksha/branch-A-JavaDoc

Add JavaDoc comments

roshnidaksha committed:
Refine user messages for better clarity

roshnidaksha committed:
Add user guide with instructions

roshnidaksha committed:
Update message display

roshnidaksha committed:
fix(Event): add validation to ensure event start time is before end time

roshnidaksha committed:
fix: save data immediately after add and delete commands

Tasks were only saved when the user exited with 'bye', which could
cause data loss if the program crashed.

Let's save tasks immediately after every add or delete command to
ensure data persistence.

roshnidaksha committed:
Add command to clear all tasks and update README.md

roshnidaksha committed:
Update README.md to improve formatting

roshnidaksha committed:
fix(Parser): remove unnecessary checks

roshnidaksha committed:
fix(Parser): fix typo in Parser regex

roshnidaksha committed:
Update Javadoc to sync with date format

roshnidaksha committed:
fix(DateParser): make date parser cross-platform

okkhoy commented on src/main/java/planit/Planit.java:
```diff
@@ -0,0 +1,39 @@
+        while (!exitPlanit) {
+            Ui.showPlanitUserGuide();
+            String userChoice = Ui.getUserInput();
+            if (userChoice.equalsIgnoreCase("planit") || userChoice.equals("1")) {
+                taskManager.run();
+            } else if (userChoice.equalsIgnoreCase("play") || userChoice.equals("2")) {
+                Echo.echo();
+            } else if (userChoice.equalsIgnoreCase("bye") || userChoice.equals("3")) {
+                exitPlanit = true;
+            } else {
+                Ui.showError(PlanitMessages.PLANIT_INVALID_OPTION);
```
in your tP we would like to see one-shot commands rather than menu-based commands like this

okkhoy commented on src/main/java/planit/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,66 @@
+            feedback.add(newTask.toString());
+            feedback.add(String.format(PlanitMessages.TASK_LIST_SIZE, tasks.taskCount));
+            tasks.saveTasks();
+        } catch (Exception e) {
```
catching Exception is too generic. consider adding some custom exceptions for specific purposes.

okkhoy commented on src/main/java/planit/handler/Parser.java:
```diff
@@ -0,0 +1,138 @@
+package planit.handler;
```
`parser` may be a better name here for the package

okkhoy commented on src/main/java/planit/handler/Parser.java:
```diff
@@ -0,0 +1,138 @@
+/**
+ * Parses uses input.
+ */
+public class Parser {
```
could be `CommandParser`

okkhoy commented on src/main/java/planit/task/TaskList.java:
```diff
@@ -0,0 +1,181 @@
+     */
+    public HashMap<String, ArrayList<Task>> getTasksOnDate(String date) {
+        HashMap<String, ArrayList<Task>> tasksOnDateString = new HashMap<>();
+        for (String taskType : tasksMap.keySet()) {
+            tasksOnDateString.put(taskType, new ArrayList<>());
+            for (Task task : tasksMap.get(taskType)) {
+                if (task.toFileFormat().contains(date)) {
+                    tasksOnDateString.get(taskType).add(task);
+                }
+            }
+            if (tasksOnDateString.get(taskType).isEmpty()) {
+                tasksOnDateString.remove(taskType);
+            }
+        }
```
nesting is deep; can you think of ways to reduce it?

okkhoy reviewed (commented):
@roshnidaksha 
very well written and code, kudos!
i have pointed a few things in the comments; please take a look

roshnidaksha committed:
Refactor package and improve code readability

- Renamed package `handler` to `parser` for better clarity
- Replaced general exception handling with specific exceptions
- Reduced nesting in `find` and `date` commands by using streams

roshnidaksha committed:
fix(AddCommand): throw exception if duplicate task is added
